lib/posix/pipe.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <posix/unistd.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <sys/fdtable.h>
+
+struct pipe_fd_object {
+	struct stat stat;
+	/* counterpart file descriptor */
+	int other;
```

Comment:
Let me ask first - does `k_pipe` have a semaphore or mutex? Is it possible to notify a thread that is waiting on it? I haven't gotten there yet, so just thought it would be useful to consider a situation where I need to check whether a fd is valid.

Reply:
> Let me ask first - does k_pipe have a semaphore or mutex?

I'm not author of k_pipe, and don't keep entire picture of Zephyr kernel in mind, so to answer that, would need to consult docs and/or source, which hopefully you can get to earlier than me (working on sth else now).

> Is it possible to notify a thread that is waiting on it? I haven't gotten there yet, so just thought it would be useful to consider a situation where I need to check whether a fd is valid.

Not sure how that would be related to check whether a fd is valid. You'd definitely need to block on an empty k_pipe in POSIX read() implementation. And for robust impl, would apparently need to cancel such a read in case write end is closed. Both are possible with k_fifo (which socket code uses), not sure about k_pipe, would need to check docs.

Reply:
Hmm... I might need to look into k_fifo then. k_pipe definitely is a little wonky in its API. E.g. K_FOREVER and min_xfer == 1 block indefinitely, even when min_xfer can be satisfied.